### ISSUE TITLE
Checking values assigned inside if with shortcircuiting fails with UndefinedVariable and UnusedVariable

### ISSUE THREAD

**Issue by jnvsor** [vimeo/psalm#8522]
https://psalm.dev/r/0d3fcf03f5

With a fairly common pattern:

```php
if ($v = t() && $v instanceof B) {}
```

Psalm doesn't recognize the assignment ensures that `$v` will be defined by the time the `instanceof` test is done, which leads to paradoxical `UndefinedVariable` and `UnusedVariable` errors for the same variable.

PHP conditional expressions shortcircuit so the code above is equivalent to:

```php
if ($v = t()) {
  if ($v instanceof B) {}
}
```

This passes in psalm, but the former doesn't

**Comment by orklah** [comment]
Assignations in conditionals are hard to handle according to Matt, this is a valid issue so we'll keep this open but don't keep your hopes up on that one...

**Comment by jnvsor** [comment]
@orklah I was wrong psalm was right... Stupid me...


`if ($v = t() && $v instanceof B)` is being parsed by PHP as `if ($v = (t() && $v instanceof B))` and psalm doesn't throw an error when I write `if (($v = t()) && $v instanceof B)`

Now please excuse me while I look for all the code where I did that...
